File: src/test/java/org/vanautrui/languages/compiler/parsing/astnodes/typenodes/SubroutineTypeNodeTest.java
package org.vanautrui.languages.compiler.parsing.astnodes.typenodes;

import org.junit.Test;
import org.vanautrui.languages.compiler.lexing.Lexer;
import org.vanautrui.languages.compiler.lexing.utils.TokenList;

import static org.junit.Assert.assertEquals;

public class SubroutineTypeNodeTest {

  @Test
  public void test_subroutine_type_parsing_subroutine_with_side_effects()throws Exception{
    final String source="(PInt,MyType)~>PInt";
    Lexer lexer = new Lexer();
    TokenList list = lexer.lexCodeTestMode(source);

    TypeNode node = new TypeNode(list);
    assertEquals(source,node.getTypeName());
    assertEquals(0,list.size());
  }

  @Test
  public void test_subroutine_type_parsing_subroutine_without_side_effects()throws Exception{
    Lexer lexer = new Lexer();
    final String source="(PInt)->PInt";
    TokenList list = lexer.lexCodeTestMode(source);

    TypeNode node = new TypeNode(list);
    assertEquals(source,node.getTypeName());
    assertEquals(0,list.size());
  }

  @Test
  public void test_subroutine_type_parsing_subroutine_with_subroutine_argument()throws Exception{
    final String source="(PInt,(PInt)->PInt)->PInt";
    Lexer lexer = new Lexer();
    TokenList list = lexer.lexCodeTestMode(source);

    TypeNode node = new TypeNode(list);
    assertEquals(source,node.getTypeName());
    assertEquals(0,list.size());
  }
}

File: dragon-parser/src/test/astnodes/typenodes/SubroutineTypeNodeTest.cpp
package org.vanautrui.languages.test.astnodes.typenodes;

import org.vanautrui.languages.commandline.ParserPhases;
import org.vanautrui.languages.compiler.lexing.utils.TokenList;
import org.vanautrui.languages.compiler.parsing.astnodes.typenodes.basic_and_wrapped.SubroutineTypeNode;

public final class SubroutineTypeNodeTest {

	public static int test_typename() throws Exception {
		final String source = "(PInt,PInt)->PInt ";
		final TokenList list = ParserPhases.makeTokenList(source, false);

		final SubroutineTypeNode sub = new SubroutineTypeNode(list);

		boolean assert1 = (0 == list.size());

		return (assert1)?1:0;
	}

	public static int test_typename_subroutine_return_type() throws Exception {
		final String source = "(PInt,PInt)->((PInt)->PInt) ";
		final TokenList list = ParserPhases.makeTokenList(source, false);

		final SubroutineTypeNode sub = new SubroutineTypeNode(list);

		boolean assert1 = (0 == list.size());
		return (assert1)?1:0;
	}

	public static int test_subroutine_type_parsing_subroutine_with_side_effects() throws Exception {
		final String source = "(PInt,MyType)~>PInt ";

		final TokenList list = ParserPhases.makeTokenList(source, false);

		final SubroutineTypeNode node = new SubroutineTypeNode(list);
		boolean assert1 = (0 == list.size());
		return (assert1)?1:0;
	}

	public static int test_subroutine_type_parsing_subroutine_without_side_effects() throws Exception {

		final String source = "(PInt)->PInt ";
		final TokenList list = ParserPhases.makeTokenList(source, false);

		SubroutineTypeNode node = new SubroutineTypeNode(list);
		boolean assert1 = (0 == list.size());
		return (assert1)?1:0;
	}

	public static int test_subroutine_type_parsing_subroutine_with_subroutine_argument() throws Exception {
		final String source = "(PInt,((PInt)->PInt))->PInt ";

		final TokenList list = ParserPhases.makeTokenList(source, false);

		SubroutineTypeNode node = new SubroutineTypeNode(list);
		boolean assert1 = (0 == list.size());
		return (assert1)?1:0;
	}
}
